posts/2022-05-03-microstream-cdi-ext.adoc

```diff
@@ -0,0 +1,265 @@
+
+== CDI extension
+
+The CDI extension frees you from defining the `StorageManager` and explicitly calls the _store_ method by making use of the CDI facilities that are available with the Open Liberty server runtime.
+
+To use it, just add the MicroStream CDI extension dependency to your application.
+
+[source,XML]
+----
+	<dependency>
+		<groupId>one.microstream</groupId>
+		<artifactId>microstream-integrations-cdi</artifactId>
+		<version>07.00.00-MS-GA</version>
+	</dependency>
+----
+
+The extension currently makes use of the Embedded Storage Manager so the data is stored on disk.
+
+== Configuring StorageManager
+
+The embedded storage manager is configured through a few MicroProfile configuration values.  For those that know MicroProfile Config, as long as the values are defined in a default or configured source, the application is able to read them at startup. If you want to know more about MicroProfile Config, have a look at the https://github.com/eclipse/microprofile-config/blob/master/spec/src/main/asciidoc/microprofile-config-spec.asciidoc[specification document].
+
+All the standard MicroStream properties, as listed in the https://docs.microstream.one/manual/storage/configuration/properties.html[reference manual],  are supported using the following convention;
+
+| All dashes are replaced by . (dot) and prefixed by ‘one.microstream.’
+
+
+So the property key `one.microstream.storage.directory` refers to the _storage_directory_ property.
+
+To store the data in a certain directory on your disk, you can add the following line to the _microprofile-config.properties_ file. (absolute paths are also supported). Remember that this kind of configuration should be specified outside of your application.  Here we use the _microprofile_config.properties_ file as a simple solution.
+
+[source]
+----
+one.microstream.storage.directory=target/data
+----
+
+The storage manager is also made available as a CDI bean in case you want to access some methods programmatically. But you don't need this as we will discuss some additional functionality of the CDI extension in a moment.
+
+[source,Java]
+----
+@Inject
+Private StorageManager storageManager
+----
+
+
+== Define Root Instance
+
+As we have seen in the basic usage, we have to provide a _Root_ instance to the framework so that it can loop over the entire Object Graph and determine the instances that need to be stored and loaded at startup.
+
+Since we no longer instantiate the Storage Manager ourselves, we need a way to indicate the root instance.  This can be done by annotating a class with
+
+[source,Java]
+----
+one.microstream.integrations.cdi.types.Storage
+----
+
+This marker, and we must use it only once in our application or the dependencies of the project we have, identifies the class that denotes the root instance. An instance is automatically instantiated, defined as a CDI bean, and linked with the storage manager so that data can be persisted.
```

Comment:
```suggestion
This marker identifies the class that denotes the root instance. An instance is automatically instantiated, defined as a CDI bean, and linked with the storage manager so that data can be persisted. We can use the marker only once in our application or the dependencies of the project.
```